### PR TITLE
feat: HTTP/2 support

### PR DESCRIPTION
**Pull request by szmarczak**
**WIP**

- [x] PoC is working
- [ ] unref session when 0 pending streams, ref on stream\
- [ ] ...

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -1326,7 +1359,147 @@ function _resume (client, sync) {
...
+    if (client[kStrictContentLength]) {
+      errorRequest(client, request, new RequestContentLengthMismatchError())
+      return false
+    }
+
+    process.emitWarning(new RequestContentLengthMismatchError())
+  }
+
+  try {
+    request.onConnect((err) => {
```
Comment (szmarczak) [inline lib/client.js:1411]: Should we call `onConnect` immediately or on stream `ready` event? https://nodejs.org/api/http2.html#http2_event_ready

File: lib/client.js [hunk 2]
```diff
@@ -1326,7 +1360,153 @@ function _resume (client, sync) {
...
+  } else {
+    assert(false)
+  }
+
+  console.log('http/2 request')
+  // TODO: @szmarczak: ref only if current streams count is 0
+  session.ref()
+  const stream = session.request(headers)
+  stream.on('response', headers => {
+    if (request.onHeaders(Number(headers[':status']), headers, stream.resume.bind(stream), '') === false) {
```
Comment (szmarczak) [inline lib/client.js:1470]: The handler expects an array but the native `http2` module provides us with an object. What should we do?

Comment (ronag) [reply lib/client.js:1470]: Convert to array

File: lib/core/request.js [hunk 3]
```diff
@@ -226,7 +226,7 @@ function processHeader (request, key, val) {
   ) {
     throw new NotSupportedError('expect header not supported')
   } else {
-    request.headers += `${key}: ${val}\r\n`
+    request.headers[key] = val
```
Comment (ronag) [inline lib/core/request.js:229]: This will make http1 slower... :/

File: lib/core/connect.js [hunk 4]
```diff
@@ -29,6 +29,8 @@ function buildConnector ({ maxCachedSessions, socketPath, timeout, ...opts }) {
       const session = sessionCache.get(servername) || null
 
       socket = tls.connect({
+        // TODO: @szmarczak: move this to client.js
+        ALPNProtocols: ['h2', 'http/1.1'],
```
Comment (ronag) [inline lib/core/connect.js:33]: experimental + disabled by default?

**Comment by ronag** [comment]
Maybe would be better to target @jasnell's node quic branch? I'm not that comfortable with the node core http2 implementation.

**Comment by espoal** [comment]
@ronag what's the issue with the core node HTTP/2?

**Comment by ronag** [comment]
> @ronag what's the issue with the core node HTTP/2?

https://github.com/nodejs/node/issues?q=is%3Aopen+is%3Aissue+label%3Ahttp2

**Comment by nickmccurdy** [comment]
Superseded by #2061

**Pull request closed without merging**